aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java

```diff
@@ -165,7 +167,14 @@ public synchronized Completable start(Action onLocalStorageReady) {
                 LOG.debug("Orchestrator started.");
                 announceRemoteSyncStarted();
             })
-        ).doFinally(startStopSemaphore::release);
+        )
+        .doOnError(error -> {
+            LOG.warn("Orchestrator failed to start.", error);
+            startStopSemaphore.release();
+            //noinspection ResultOfMethodCallIgnored
+            stop().blockingAwait(ACQUIRE_PERMIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
```

Comment:
If we don't have a clean start, we should teardown again, so we're back to a deterministic stop state.

Reply:
This makes sense. Only concern I have is similar to what I mentioned earlier. If a connection is not available what the impact on calls to the DataStore APIs?

Reply:
> If a connection is not available what the impact on calls to the DataStore APIs?

Thanks for bringing this case to my attention. This is an indeed an issue:
1. Device is offline;
2. User is not signed in;
3. A call to `save(...)` is made;
4. No response is rendered until `initializeOrchestrator()` fails.
5. At that point, the `onResult` consumer _is_ triggered as desired, indicating successful use of the local storage.

(4) is a problem.

Reply:
To build on my response to the other comment, I think we only need to block interactions with the local DataStore until the StorageObserver is ready.

Maybe we need to account for more than just started/stopped for the orchestrator. 

```
if Orchestrator is STOPPED => // need to start everything and signal when StorageObserver is ready.
   triggerStartAndWaitForStorageObserver
   if StorageObserver STARTED =>
      emit storageObserverReady //This unlocks callers
   else
      Set Orchestrator back to STOPPED
      throw error back to caller.      
   [MutationProcessor, SubscriptionProcessor, SyncProcessor].asyncStart 
   if [MutationProcessor, SubscriptionProcessor, SyncProcessor] STARTED
     set Orchestrator STARTED
   else
     set Orchestrator FAILED

if STARTED => nothing to do, it's ready to go

if FAILED => trigger non-blocking retry.
   [MutationProcessor, SubscriptionProcessor, SyncProcessor].asyncStart 
   if [MutationProcessor, SubscriptionProcessor, SyncProcessor] STARTED
     set Orchestrator STARTED
   else
     set Orchestrator FAILED
```

Reply:
Resolution: I decoupled the initialization of the API synchronization code, from the code which observes the storage. Now, there is always a storage observer when `beforeOperation()` completes, and so any customer-provided data will be observed. The API sync may start at a later time.